druid/src/widget/focus.rs

```diff
@@ -0,0 +1,232 @@
+// Copyright 2020 The druid Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A focus widget.
+
+use druid::widget::prelude::*;
+
+use druid::{commands, Data, FocusNode, HotKey, KbKey, Point, Rect, SysMods, Widget, WidgetPod};
+
+#[derive(Debug, Clone)]
+pub(crate) struct FocusController {
+    focus_node: FocusNode,
+    previous_focus_node: FocusNode,
+    requested_focus: bool,
+    pub(crate) auto_focus: bool,
+}
+
+impl FocusController {
+    pub(crate) fn new() -> Self {
+        Self {
+            focus_node: FocusNode::empty(),
+            previous_focus_node: FocusNode::empty(),
+            requested_focus: false,
+            auto_focus: false,
+        }
+    }
+
+    pub(crate) fn event_start(&mut self, ctx: &mut EventCtx) {
+        self.previous_focus_node = ctx.focus_node();
+        ctx.set_focus_node(self.focus_node);
```

Comment:
I'm still confused about this, and I'm giving up on trying to figure it out from the code. What happens if we don't set this? What things will break?

Reply:
Firstly lets me explain what `FocusNode` actually is and how I use it. The main idea is to pass information about the `Focus` widget state to its descendants. Right now, `is_focused` shows if the current widget has a focus and hence `is_focused = false` for its descendants. In the example, we can draw decoration over the `Button` just because we have `focus_node` information that depicts the `Focus` widget state. Also, descendants of the `Focus` widget should also be able to receive keyboard events, that's why we should set the current `focus_node` to the context, to check if descendants can receive events. In code I have added this: `ctx.state.focus_node.is_focused || self.state.has_focus` to check if widget can receive keyboard events.

And here is a little diagram:

```
"Focus" <- Owner of `focus_node` put it to the context; `is_focused = true`
|--"Widget 1" <- Can read `focus_node` by calling `ctx.focus_node()`; `is_focused = false` but `ctx.focus_node().is_focused = true`
|--|--"Widget 2" <- The same properties as Widget 1
|--"Widget 3" <- The same properties as Widget 1
```

Reply:
okay, that's a good explanation, thanks. I think it would make sense to write something like that down in the docs for `FocusNode`?

Reply:
Okay, I'll try )

Reply:
@cmyr what about this:
```
/// A type that depicts the current [`Focus`] widget state for its descendants.
///
/// This type can be retrieved from any descendants of [`Focus`].
/// For example, this type can be used to draw focus decoration by reading the current is_focused
/// value that shows if [`Focus`] widget is_focused. Unlike ctx.is_focused, ctx.focus_node().is_focused
/// will be true even if the current widget is not focused and if the parent [`Focus`] has focus.
///
/// This type also allows descendants to receive keyboard events if the parent [`Focus`] widget has focus.
```